certbot-dns-bookmyname/certbot_dns_bookmyname/dns_bookmyname.py

```diff
@@ -0,0 +1,330 @@
+"""DNS Authenticator for BookMyName."""
+import logging
+
+import zope.interface
+
+from certbot import errors
+from certbot import interfaces
+from certbot.plugins import dns_common
+
+import requests
+from bs4 import BeautifulSoup
+import re
+
+logger = logging.getLogger(__name__)
+
+"""
+BookMyName account manager
+
+NOTE:
+
+There is an API provided by BookMyName to big (500+ domains) accounts.
+This file is NOT a client implementation of this BookMyName API.
+This file is just a wrapper around simple HTTP website access.
+
+The manager is not unit-testable as it requires a BookMyName account
+and there is no such test account available.
+
+OTOH, for unit-testing the _BookMyNameClient class, a mockup of this
+manager is used, which simulates an account and a domain.
+"""
+
+class Error(Exception):
+    """Generic BookMyName error"""
+
+class LoginError(Error):
+    """BookMyname login failure"""
+
+class DomainError(Error):
+    """BookMyName domain access failure"""
+
+class ZoneError(Error):
+    """BookMyName zone access failure"""
+
+class Manager(object):
+    """BookMyName account manager
+
+    This class handles the managing of a BookMyName account.
+    It allows logging in, listing the domains registered to this account
+    and reading and writing the zone of a registered domain.
+    """
+
+    description = 'Manages a BookMyName DNS account.'
+
+    def __init__(self, login, password, http_session):
+        self.login = login
+        self.password = password
+        self.session = http_session
+        self.logged_in = False
+
+    def log_in(self):
+        """
+        Log in on the website
+
+        Open the website login page and pass the handle (login/id) and
+        password. The session identifier will remain in self.session
+        and allow further POST and GET requests to succeed.
+
+        This method can be called several times; it will only log in
+        once.
+        """
+        if not self.logged_in:
+            try:
+                self.session.post(
+                    "https://www.bookmyname.com/login.cgi",
+                    {
+                        'handle': self.login,
+                        'passwd': self.password
+                    }
+                )
+                self.logged_in = True
+            except:
+                raise LoginError
+
+    def get_domains_and_ids(self):
+        """
+        Get the list of registered domains with their corresponding IDs
+
+        In order to get the domains and IDs, we need to parse the table
+        of domains. Each domain anchor in the first column provides the
+        name (in the inner HTML) and the ID (at the end of the href URI)
+
+        WATCH OUT: this has not been tested against cases where not all
```

Comment:
This scenario concerns me, especially because of the short validity of Let's Encrypt's certificates.

Someone could begin using this plugin with a set of domains, configure a cron job for certbot to renew their certificates automatically, and then register additional domains with BookMyName causing some domains to be pushed onto a second page, at which point some certificates could no longer be renewed.

How hard would it be to handle paging?

Reply:
The main problem is that I don't have enough BMN domains myself there to be able to tell how multiple domain pages actually look like (or even if they do appear), and to test the plugin against those cases. I'll try and look for someone who has and is willing to help, but I can't make promises.

Reply:
FYI, I've obtained from the BookMyName support desk a way to really get all domains.

Reply:
New PR uses a URI which returns all domains in a CSV list (names, not IDs, but IDs are now unnecessary)